### ISSUE TITLE
브랜치 전략 : feature-data-eda-loader

### ISSUE THREAD

**Issue by iejob** [iejob/computer-vision-competition-1SEN#6]
## 목표와 범위
- **목표:**
    학습·검증·추론에 공통 사용 가능한 **안전하고 강건한 DataLoader 파이프라인**(+ EDA 리포트) 완성.
    
    - **데이터 무결성 검증, 정량 EDA, 학습/검증/추론 공통 로더·변환·샘플러 구축**을 일원화하여 모델 개발 기반 제공.
    - 모든 기능은 **`ipynb`에서 사전 검증** 후, 통과 코드만 `src/` 표준 디렉터리로 **모듈화 이관**.
    - 병합 전 **재현성·성능(처리량)·안전성(규칙 준수)** 체크리스트 충족을 완료 기준(DoD)으로 정의.
- **범위:** 데이터 스키마/무결성 검사, 클래스 분포/불균형 대응, 문서 이미지 특화 증강, 빠른 로딩·캐싱, 샘플러(WeightedRandomSampler)·스플릿(누수 방지)·에러 내성 I/O.
    - **데이터 거버넌스:** 스키마/라벨/파일 무결성, 중복·누수 검출, 규칙 준수 자동 점검.
    - **로더/트랜스폼/샘플러:** 문서 특화 전처리·증강, 가중 샘플러, Stratified KFold 유틸.
    - **EDA 리포팅:** 분포/품질/증강 효과 시각화 + HTML/PDF 리포트.
    - **성능·신뢰성:** I/O 안정성, 처리량 벤치마크, 재현성(시드/버전/환경 로그).

---

## 현재 데이터·환경 요약(정량 근거)
- `train.csv` **1570행 × 2컬럼**(`ID`,`target`) — **고유 ID=1570**
- `meta.csv` **17행 × 2컬럼**(`target`,`class_name`) — 0~16 매핑 완비
- `sample_submission.csv` **3140행 × 2컬럼**(`ID`,`target`) — **고유 ID=3140**
- **클래스 분포(정렬)**:
    
    `0:100, 1:46, 2:100, 3:100, 4:100, 5:100, 6:100, 7:100, 8:100, 9:100, 10:100, 11:100, 12:100, 13:74, 14:50, 15:100, 16:100`
    
- **권고 클래스 가중치(중앙값=100 기준)**:
    
    ```
    {0:1.0, 1:2.1739, 2:1.0, 3:1.0, 4:1.0, 5:1.0, 6:1.0, 7:1.0,
     8:1.0, 9:1.0, 10:1.0, 11:1.0, 12:1.0, 13:1.3514, 14:2.0, 15:1.0, 16:1.0}
    ```
    
- **요구 라이브러리 확인:** `requirements.txt` 기준 `albumentations==1.3.1, pandas==2.1.4, Pillow==9.4.0, timm==0.9.12` 등 기재.
    
    > 권고: 플랫폼 종속성으로 인해 torch/torchvision/torchaudio는 별도 설치·핀 고정(예: CUDA 12.x 환경에 맞게) — 브랜치 문서에 버전 정책을 명시하십시오.
    > 

---

## 구성·설정 단일 소스(추가)
**목적:** 경로/파라미터/정책을 한 곳에서 관리하여 재현성·감사를 담보.

**`configs/data_config.json` (예시)**

```json
{
  "paths": {
    "train_csv": "data/raw/train.csv",
    "test_csv": "data/raw/sample_submission.csv",
    "meta_csv": "data/raw/meta.csv",
    "train_root": "data/raw/train",
    "test_root": "data/raw/test",
    "processed_dir": "data/processed",
    "reports_dir": "docs",
    "exp_dir": "experiments/data_loader",
    "logs_dir": "logs"
  },
  "loader": {
    "img_size": 224,
    "batch_size": 64,
    "num_workers": 4,
    "pin_memory": true,
    "prefetch_factor": 2,
    "persistent_workers": true
  },
  "augs": {
    "rotate_limit": 25,
    "jitter": 0.2,
    "use_clahe": true,
    "use_blur": true,
    "use_affine": true,
    "hflip_p": 0.2
  },
  "sampler": { "use_weighted": true, "clip_min": 0.5, "clip_max": 3.0 },
  "seed": 42,
  "strict_mode": true
}
```

> strict_mode: 무결성 위반 시 즉시 실패(머지 전 필수); 탐색 단계는 false 허용.
> 

---

## 파라미터 표준(요약·상한 가드)
| 단계 | 파라미터 | 기본값 | 허용 범위 | 비고 |
| --- | --- | --- | --- | --- |
| Resize/Pad | size | 224 | {224,320,384} | Val/Test 동일 |
| Rotate | limit(°) | ±25 | ±15~±30 | p ∈ [0.5,0.8] |
| RandomResizedCrop | scale | (0.8,1.0) | (0.7,1.0) | ratio 0.9~1.1 |
| HorizontalFlip | p | 0.2 | 0.0~0.3 | 문서 특성상 과다 금지 |
| ColorJitter | strength | 0.2 | 0.1~0.3 | 밝기/대비/채도 |
| CLAHE | clip_limit | 2.0 | 1.0~3.0 | p ≤ 0.3 권장 |
| GaussianBlur | k | 3~5 | 3~5 | p ≤ 0.3 권장 |
| Affine | shear/scale | 소폭 | 제한적 | 기하 왜곡 최소화 |
| Normalize | mean/std | ImageNet | 고정 | timm 호환 |
| ToTensorV2 | - | - | - | albumentations.pytorch |

> 위 표의 기본값 = 코드 default로 통제하고, 변경은 data_config.json에서만.
> 

---

## 실패 정책(Failure Policy, 추가)
- **파일 누락/손상**
    - *Strict*: 즉시 실패 → `logs/data_checks.log`, `data/processed/data_issues.csv` 기록 후 수정 재실행.
    - *Lenient*: 블랭크 대체/샘플 스킵 + 경고 로그(탐색용, 머지 금지).
- **CSV 스키마 불일치**: *항상 Strict* — 필수 컬럼·라벨 도메인 미일치 시 실패.
- **중복/누수 발견**: *항상 Strict* — 제거/마스킹 목록 확정 후 재검증(0건이 되도록).

---

### 1) 데이터 스키마 & 무결성 점검
- **필수 검사**:
    - `train.csv`(`ID`,`target`) 스키마·도메인, `meta.csv` 0~16 매핑 완비, `sample_submission.csv` 길이=테스트 이미지 수·ID 집합 일치.
    - **파일 존재**: `data/raw/train|test/*.jpg` 전수 검사(미존재/0바이트/오픈 실패 로그).
    - **중복/누수**: 동일 `ID`/MD5 중복, **train-test 교차 중복=0**.
- **결과 산출**: `data/processed/data_schema.json`, `data/processed/data_issues.csv`(발견 시), `logs/data_checks.log`.

### 2) 데이터셋 & 변환(Albumentations)
- **`DocTypeDataset`** (모드: `train|val|test`)
    - 입력: `csv_path`, `img_root`, `mode`, `transform`, `return_id`
    - 출력: `(image_tensor, label)` 또는 `(image_tensor, id)`
    - **예외 내성**: 로딩 실패 리트라이 후 **대체 샘플**(평균 픽셀) 반환 + 로그.
- **전처리/증강 파이프라인**
    - 공통: `LongestMaxSize → PadIfNeeded → Normalize(Imagenet) → ToTensorV2`
    - Train: Rotate/RandomResizedCrop/HFlip 제한/ColorJitter/CLAHE/GaussianBlur/Sharpen/Affine(소폭)
    - Val/Test: **과증강 금지**(Resize/Pad + Normalize)
    - **옵션 훅**: CutMix/MixUp 인터페이스 제공(라벨 믹싱은 학습 루프에서 처리).

### 3) 샘플러 & 스플릿
- **WeightedRandomSampler**: `w_c = clip(median(count)/count_c, [0.5,3.0])`
    - 현재 권고: `1→2.1739, 13→1.3514, 14→2.0`, 나머지 1.0
- **Stratified Split 유틸**: `n_splits=5, shuffle=True, random_state=42`
- **배치 빌더**: `DataLoader` 튜닝 파라미터 옵션화(`num_workers/pin_memory/prefetch_factor/persistent_workers`)

### 4) 시각화 & EDA
- 클래스 분포/해상도·종횡비/밝기·대비 히스토그램, **증강 전·후 패널**(클래스별).
- **쿼런틴 뷰어**: 로딩 실패·의심 이미지 미리보기.
- **리포트**: HTML/PDF 내보내기(버전/시드/구성 요약 포함).

### 5) 성능/안정화
- **I/O 최적화**: Pillow-SIMD(가능 시), JPEG 디코더 예외 처리 통일, 썸네일 캐시(옵션).
- **재현성**: 시드 고정, worker 시드 설정(`worker_init_fn`), 라이브러리 버전, 하드웨어 요약을 `config.json` 저장.

---

## ipynb-우선 워크플로우 (파일명·셀 체크리스트)
1. **`notebooks/EDA/00_data_schema_and_integrity.ipynb`**
    - [ ]  CSV 스키마·도메인 검증 → `data_schema.json`/`data_issues.csv`
    - [ ]  파일 존재/해시 중복/누수 검사 → `logs/data_checks.log`
    - [ ]  클래스 분포/불균형 요약
2. **`notebooks/EDA/01_transforms_gallery.ipynb`**
    - [ ]  Train/Val/Test 파이프라인 정의
    - [ ]  증강 전/후 시각화(5~10 클래스)
3. **`notebooks/EDA/02_dataloader_smoke_test.ipynb`**
    - [ ]  Dataset/DataLoader 배치 점검(shape/값 범위/정규화)
    - [ ]  `num_workers`/`batch_size` 스윕 → **imgs/sec** 로그
4. **`notebooks/EDA/03_weighted_sampler_sanity.ipynb`**
    - [ ]  10k 샘플 시뮬 → 클래스 편차 ≤ ±8%p
5. **`notebooks/EDA/04_split_and_leakage_checks.ipynb`**
    - [ ]  Stratified KFold 분할 → fold별 분포 표/누수 0
6. **`notebooks/EDA/05_export_to_src.ipynb`**
    - [ ]  통과 코드만 `src/data/*.py`, `src/utils/*.py` 이관
    - [ ]  `__all__`/docstring/타입힌트 정리

> 자동 실행(권고): papermill 또는 nbconvert --execute로 00~05 일괄 실행 및 산출물 검증.
> 

---

## 산출물(Artifacts)
- **코드 모듈**
    - `src/data/dataset.py` : `DocTypeDataset`
    - `src/data/transforms.py` : `get_train_transforms()`, `get_val_transforms()`
    - `src/data/samplers.py` : 클래스 가중치 계산·샘플러 생성
    - `src/utils/splits.py` : Stratified KFold 유틸
    - `src/utils/checks.py` : 스키마/파일/해시 중복/누수 점검
    - `src/utils/image_io.py` : 예외 내성 I/O(리트라이·블랭크 대체·썸네일옵션)
- **리포트/로그**
    - `docs/eda_report.html` (또는 `.pdf`)
    - `experiments/data_loader/<run_id>/{config.json, timings.json, data_checks.log}`
    - `data/processed/{data_schema.json, class_weights.json, folds_k5_seed42.csv}`
- **노트북(최종 파일명 확정)**
    - `notebooks/EDA/00_data_schema_and_integrity.ipynb`
    - `notebooks/EDA/01_transforms_gallery.ipynb`
    - `notebooks/EDA/02_dataloader_smoke_test.ipynb`
    - `notebooks/EDA/03_weighted_sampler_sanity.ipynb`
    - `notebooks/EDA/04_split_and_leakage_checks.ipynb`
    - `notebooks/EDA/05_export_to_src.ipynb`

---

### 1. 데이터 스키마·무결성
- `train.csv` 필수 컬럼·라벨 도메인·결측 0, `meta.csv` 전 범위 매핑, `sample_submission.csv`와 테스트 ID 집합 일치.
- 파일 존재/0바이트/오픈 실패 0건.
- MD5/CRC32 기반 **Train-Test 교차 중복 0**, Train 내 중복 0(불가피 시 제외 목록 확정 후 재검증).
- 결과 기록: `data_checks.log`, `data_issues.csv`, `data_schema.json`.

### 2. 데이터셋·변환(문서 특화)
- `DocTypeDataset`(`train|val|test`), 실패 내성(리트라이·대체 샘플)·로깅.
- Albumentations 파이프라인(상단 표준 범위 준수).
- CutMix/MixUp 훅 인터페이스 제공(적용은 학습 루프).

### 3. 샘플러·스플릿
- WeightedRandomSampler(중앙값/빈도 비례·클리핑).
- StratifiedKFold: `n_splits=5, shuffle=True, random_state=42`.
- DataLoader 튜닝 파라미터 옵션화.

### 4. EDA·시각화
- 분포/해상도·종횡비/밝기·대비, 증강 전·후 패널, 쿼런틴 뷰어.
- HTML/PDF 리포트(버전·시드·파이프라인 요약 포함).

### 5. 성능·신뢰성
- 처리량(imgs/sec)·배치 메모리 사용량 측정.
- Pillow-SIMD 옵션, 디코더 예외 처리 일원화.
- 재현성: 파라미터·버전·시드·하드웨어 요약을 `config.json` 저장.

---

## 실험·검증(Validation) 기준
| 항목 | 측정/검증 방법 | 통과 기준(제안) |
| --- | --- | --- |
| **스키마/무결성** | 00 노트북 로그/CSV | 에러 0건(재실행 시 0건) |
| **중복/누수** | MD5 교집합표 | 교차 0건, 내부 0건(제외 목록 확정 시 OK) |
| **증강 유효성** | 01 갤러리 + 체크리스트 | 텍스트/레이아웃 보존, 과도 변형 없음 |
| **로더 스모크** | 02 100배치 | 예외 0건, 텐서 shape/정규화 정상 |
| **샘플러 균형** | 03 10k 시뮬 | 클래스 편차 ≤ **±8%p** |
| **스플릿 누수** | 04 fold 교차 | ID/해시 교차 0건 |
| **처리량 벤치마크** | 02 imgs/sec | 224px, `bs=64`, `num_workers≥4`에서 팀 합의 **≥ X imgs/s** |
| **재현성** | 동일 환경 재실행 | 로그 지표 ±1% 이내 |

> X imgs/s는 최초 측정 후 문서에 고정(하드웨어·설정 표기).
> 

---

## 병합 조건(Definition of Done & PR 체크리스트)
- **DoD**
    - ✅ 00~05 노트북 **모든 셀 실행 완료(오류·중대 워닝 0)**
    - ✅ `docs/eda_report.html|pdf`, `experiments/.../{config.json,timings.json,data_checks.log}` 산출
    - ✅ `src/` 모듈 이관(타입힌트·독스트링·예외 처리 포함)
    - ✅ 재현성·성능·안전성 기준 충족(상기 표)
    - ✅ 규칙 준수(외부 데이터/PL/유료 API 미사용) 체크
- **PR 체크리스트**
    - [ ]  코드 포맷팅(black/isort), 린트 통과
    - [ ]  단위 테스트 3종 통과(스키마/샘플러/트랜스폼)
    - [ ]  README 갱신(사용법·옵션·벤치마크 표)
    - [ ]  아티팩트 링크(리포트·로그) 첨부
    - [ ]  **(환경)** PyTorch/torchvision 버전 정책 명시(설치 메모 포함)

---

## 브랜치 작업 순서(권장 커밋 플로우)
1. `feat(checks)`: 스키마·파일 무결성·중복/누수 점검(00)
2. `feat(transforms)`: 전처리/증강 정의·갤러리(01)
3. `feat(dataloader)`: Dataset/DataLoader 스모크·성능 측정(02)
4. `feat(sampler)`: 가중 샘플러 도입·시뮬(03)
5. `feat(splits)`: Stratified KFold·누수 점검(04)
6. `feat(export)`: `src/` 이관 + 실험 설정/로그(05)
7. `docs(eda)`: 리포트/README 업데이트
8. `test(unit)`: 최소 단위 테스트 3종 추가

---

## 유닛 테스트(명시적 파일명, 추가)
- `tests/test_checks.py`
    - 스키마·라벨 도메인·결측 0 / 파일 누락·0바이트·오픈 실패 0
- `tests/test_sampler.py`
    - 가중치 키(0~16)·양수 검증 / 10k 시뮬 분포 편차 ≤ ±8%p
- `tests/test_transforms.py`
    - Train 파이프: 텐서 dtype=float32·정규화 범위 / Val·Test: 증강 없음 확인
- `tests/test_splits.py` *(권장)*
    - Fold 분포 근사 동일·ID/해시 교차 0

---

## 성능 벤치마크 절차(명시)
1. 224px, `bs ∈ {32,64,128}`, `num_workers ∈ {2,4,8}` 전수 스윕
2. 각 조합 500배치(워밍업 제외) 평균 imgs/sec 기록
3. 최상 조합의 **하위 10% 프랙타일**을 기준치 **X imgs/s**로 고정
4. `experiments/data_loader/<run_id>/timings.json`에 저장(하드웨어·조합·분산 포함)

---

## 비목표(Non-goals)
- OCR/LLM 파이프라인 구성(스코프 외)
- Pseudo Labeling(규칙 위반)
- 외부 데이터/사전 라벨 확장(규칙 위반)

---

## 리스크·준수사항
- **규칙 준수:** 테스트 데이터 **학습 활용 금지**, 외부 데이터/유료 API 미사용.
- **품질 리스크:** 과도 증강으로 텍스트/레이아웃 훼손 → 상한 가드·갤러리 체크리스트.
- **성능 리스크:** 고해상도·대배치로 I/O 병목/OOM → `num_workers`/`pin_memory`/`prefetch_factor` 튜닝, Pillow-SIMD 고려.
- **환경 리스크:** `torch/torchvision` 미핀 → **버전 정책 명시 및 설치 메모** 문서화(WSL/CUDA 맞춤).
